tutorials/src/bin/basic-tutorial-8.rs

```diff
@@ -0,0 +1,260 @@
+    );
+
+    let data: Arc<Mutex<CustomData>> =
+        Arc::new(Mutex::new(CustomData::new(appsrc.clone(), appsink.clone())));
+
+    let data_clone = data.clone();
+    appsrc.connect_need_data(move |_, _size| {
+        let data = &data_clone;
+        let mut d = data_clone.lock().unwrap();
+
+        if d.source_id.is_none() {
+            println!("start feeding");
+
+            let data_clone = data.clone();
+            d.source_id = Some(idle_add(move || {
+                let data = &data_clone;
+                let mut samples: Vec<u8> = vec![0; CHUNK_SIZE];
+                let num_samples = CHUNK_SIZE / 2; /* Each sample is 16 bits */
+                let appsrc;
+                let pts;
+
+                {
+                    let mut data = data.lock().unwrap();
+                    let samples = samples.as_mut_slice_of::<i16>().unwrap();
+                    pts = data.num_samples
+                        .mul_div_floor(gst::SECOND, SAMPLE_RATE as u64)
+                        .unwrap();
+
+                    // Generate some psychodelic waveforms
+                    data.c += data.d;
+                    data.d -= data.c / 1000.0;
+                    let freq = 1100.0 + 1000.0 * data.d;
+
+                    for sample in samples.iter_mut() {
+                        data.a += data.b;
+                        data.b -= data.a / freq;
+                        *sample = 500 * (data.a as i16);
+                    }
+
+                    data.num_samples += num_samples as u64;
+                    appsrc = data.appsrc.clone();
+                }
+
+                let mut buffer = gst::Buffer::from_mut_slice(samples).unwrap();
+                {
+                    let buffer = buffer.get_mut().unwrap();
+
+                    buffer.set_pts(pts);
+                    buffer.set_duration(
+                        (CHUNK_SIZE as u64)
+                            .mul_div_floor(gst::SECOND, SAMPLE_RATE as u64)
+                            .unwrap(),
+                    );
+                }
+
+                match appsrc.push_buffer(buffer) {
+                    gst::FlowReturn::Ok => glib::Continue(true),
+                    _ => glib::Continue(false),
+                }
+            }));
+        }
+    });
+
+    let data_clone = data.clone();
+    appsrc.connect_enough_data(move |_| {
+        let data = &data_clone;
+
+        let mut data = data.lock().unwrap();
+        if let Some(source) = data.source_id.take() {
+            println!("stop feeding");
+            source_remove(source);
+        }
+    });
+
+    // configure appsink
+    app_sink.set_property("emit-signals", &true).unwrap();
+    app_sink.set_property("caps", &audio_caps).unwrap();
+
+    let data_clone = data.clone();
+    appsink.connect_new_sample(move |_| {
+        let appsink;
+        {
+            let data = &data_clone.lock().unwrap();
+            appsink = data.appsink.clone();
+        }
```

Comment:
Also here, let the block return the appsink clone

Reply:
done.